utils/treesitter/grammar.js

```diff
@@ -0,0 +1,507 @@
+
+    match_clause: ($) =>
+      seq(choice(seq('case', $.pattern), 'default'), '=>', $.block),
+
+    match_statement: ($) =>
+      seq('match', '(', $._expression, ')', '{', repeat($.match_clause), '}'),
+
+    returned_var_statement: ($) => seq('returned', $.var_declaration, ';'),
+
+    while_statement: ($) => seq('while', '(', $._expression, ')', $.block),
+
+    break_statement: ($) => seq('break', ';'),
+
+    continue_statement: ($) => seq('continue', ';'),
+
+    return_statement: ($) =>
+      seq('return', optional(choice('var', $._expression)), ';'),
+
+    if_statement: ($) =>
+      seq('if', '(', $._expression, ')', $.block, optional($.else)),
+
+    else: ($) => choice(seq('else', $.if_statement), seq('else', $.block)),
+
+    for_statement: ($) =>
+      seq(
+        'for',
+        '(',
+        $._non_expression_pattern,
+        'in',
+        $._expression,
+        ')',
+        $.block
+      ),
+
+    statement: ($) =>
+      choice(
+        seq($._expression, ';'),
+        $.assign_statement,
+        $.var_declaration,
+        $.let_declaration,
+        $.match_statement,
+        $.returned_var_statement,
+        $.if_statement,
+        $.while_statement,
+        $.break_statement,
+        $.continue_statement,
+        $.return_statement,
+        $.for_statement
+      ),
+
+    _statement_list: ($) => seq($.statement, repeat($.statement)),
+
+    block: ($) => seq('{', optional($._statement_list), '}'),
+
+    declared_name: ($) => repeat_sep1($.ident, '.'),
+
+    generic_binding: ($) =>
+      seq(optional('template'), $.ident, ':!', $._expression),
+
+    deduced_param: ($) =>
+      choice(
+        $.generic_binding,
+        seq(optional('addr'), $.ident, ':', $._expression)
+      ),
+
+    deduced_params: ($) => seq('[', comma_sep($.deduced_param), ']'),
+
+    tuple_pattern: ($) => $.paren_pattern,
+
+    return_type: ($) => seq('->', choice('auto', $._expression)),
+
+    function_declaration: ($) =>
+      seq(
+        optional(choice('abstract', 'virtual', 'impl')),
+        'fn',
+        $.declared_name,
+        optional($.deduced_params),
+        $.tuple_pattern,
+        optional($.return_type),
+        choice($.block, ';')
+      ),
+
+    namespace_declaration: ($) => seq('namespace', $.declared_name, ';'),
+
+    alias_declaration: ($) =>
+      seq('alias', $.declared_name, '=', $._expression, ';'),
+
+    type_params: ($) => $.tuple_pattern,
+
+    interface_body_item: ($) =>
+      choice(
+        $.function_declaration,
+        seq('let', $.generic_binding, ';'),
+        seq('extend', $._expression, ';'),
+        seq('require', $._expression, 'impls', $._expression, ';')
+      ),
+
+    interface_body: ($) => seq('{', repeat($.interface_body_item), '}'),
+
+    interface_declaration: ($) =>
+      seq(
+        'interface',
+        $.declared_name,
+        optional($.deduced_params),
+        optional($.type_params),
+        $.interface_body
```

Comment:
Explorer only allows an interface body here, but the design doc also allows `;` here, for forward declarations.